### PR TITLE
overlay: add option to toggle game UI

### LINKED ISSUE #1656: Feature request: allow toggling in-game UI
Add an input mapping to allow toggling the in-game UI e.g. bars, ammo text etc.

### PR DESCRIPTION
**Pull request by lahm86**
Resolves #1656.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This adds an option to allow toggling the in-game UI, by default it is bound to `H` but can be remapped.

It will hide all bars, ammo text and pickups - let me know if you think the latter is perhaps undesirable. It will also hide the healthbar when you land on medipacks in the inventory; my thinking was people may wish to use this feature for challenge runs, e.g. complete the game without knowing your health/air status. But again, LMK if this is perhaps a step too far.

### CODE REVIEW

**Comment by aredfan** [comment]
I found 2 minor issues.

- If I change the key for Toggle UI, it also makes the UI enable/disable while in controls.
- If pickups are collected when the UI is disabled, then these pickups will queue until the UI is re-enabled. This might not be an issue though if players opt for one setting or the other.

**Comment by Richard-L** [comment]
Great idea Lahm. I can also see it useful for content creators who may want to record UI-less footage for a YT video or similar.

Can see the question re pickups. Personally I would lean towards hiding them too.

**Comment by lahm86** [comment]
> I found 2 minor issues.
> 
> * If I change the key for Toggle UI, it also makes the UI enable/disable while in controls.
> * If pickups are collected when the UI is disabled, then these pickups will queue until the UI is re-enabled. This might not be an issue though if players opt for one setting or the other.

Thanks, the pickup queue issue should be fixed now. I think it's best not to queue items at all.

The issue with changing input - this is tricky. It also seems to happen if you change the console key i.e. the console opens. If you are very quick and just tap the key when changing it, it doesn't occur, but pressing it longer makes it happen.
@rr-, @walkawayy - any thoughts on the best way to handle this? My only thinking is doing a check in `Shell_ProcessInput` to disallow changing the setting if in the inventory, but I don't like that really.

**Comment by lahm86** [comment]
> Great idea Lahm. I can also see it useful for content creators who may want to record UI-less footage for a YT video or similar.
> 
> Can see the question re pickups. Personally I would lean towards hiding them too.

Thanks, credit to @rr- though for this option.

**Comment by rr-** [comment]
> > I found 2 minor issues.
> > 
> > * If I change the key for Toggle UI, it also makes the UI enable/disable while in controls.
> > * If pickups are collected when the UI is disabled, then these pickups will queue until the UI is re-enabled. This might not be an issue though if players opt for one setting or the other.
> 
> Thanks, the pickup queue issue should be fixed now. I think it's best not to queue items at all.
> 
> The issue with changing input - this is tricky. It also seems to happen if you change the console key i.e. the console opens. If you are very quick and just tap the key when changing it, it doesn't occur, but pressing it longer makes it happen. @rr-, @walkawayy - any thoughts on the best way to handle this? My only thinking is doing a check in `Shell_ProcessInput` to disallow changing the setting if in the inventory, but I don't like that really.

We can tackle this in a separate ticket, as this dialog needs a rewrite (to use the UI/ module and reuse TR2X's code) anyway.

File: src/tr1/config.h [hunk 1]
```diff
@@ -121,6 +121,7 @@ typedef struct {
         int32_t turbo_speed;
         bool pretty_pixels;
         bool enable_reflections;
+        bool enable_ui;
```
Comment (rr-) [inline src/tr1/config.h]: I think the `ui` struct would be a more suitable place to put this setting. WDYT?

**Review by aredfan — APPROVED** [review]

File: src/tr1/config.h (hunk 1 above)
Comment (lahm86) [reply src/tr1/config.h]: Absolutely. Done 👍

**Review by rr- — COMMENTED** [review]
One more observation: currently, if a player hides the UI immediately after picking up an item and then reveals it again, the pickup animation or sprite continues to be displayed. However, if they perform the actions in reverse order – hiding the UI first, picking up an item, and then showing the UI – the animation does not appear. This seems inconsistent. Here's a suggested patch to address this issue:


```diff
diff --git a/src/tr1/game/overlay.c b/src/tr1/game/overlay.c
index 46903dd2..aec02ec2 100644
--- a/src/tr1/game/overlay.c
+++ b/src/tr1/game/overlay.c
@@ -251,6 +251,9 @@ static void M_BarGetLocation(
 
 void Overlay_BarDraw(BAR_INFO *bar_info, RENDER_SCALE_REF scale_ref)
 {
+    if (!g_Config.ui.enable_ui) {
+        return;
+    }
     const RGBA_8888 rgb_bgnd = { 0, 0, 0, 255 };
     const RGBA_8888 rgb_border = { 53, 53, 53, 255 };
 
@@ -320,6 +323,9 @@ static float M_Ease(int32_t cur_frame, int32_t max_frames)
 
 static void M_DrawPickup3D(DISPLAY_PICKUP *pu)
 {
+    if (!g_Config.ui.enable_ui) {
+        return;
+    }
     int32_t screen_width = Screen_GetResWidth();
     int32_t screen_height = Screen_GetResHeight();
     int32_t pickup_width = screen_width / 8;
@@ -481,6 +487,10 @@ static void M_DrawPickupsSprites(void)
             continue;
         }
 
+        if (!g_Config.ui.enable_ui) {
+            continue;
+        }
+
         const int32_t x =
             Viewport_GetWidth() - sprite_height - sprite_width * pu->grid_x;
         const int32_t y =
@@ -587,6 +597,11 @@ static void M_RemoveAmmoText(void)
 
 static void M_DrawAmmoInfo(void)
 {
+    if (!g_Config.ui.enable_ui) {
+        M_RemoveAmmoText();
+        return;
+    }
+
     const double scale = 1.5;
     const int32_t text_height = 17 * scale;
     const int32_t text_offset_x = 3;
@@ -715,10 +730,6 @@ void Overlay_HideGameInfo(void)
 void Overlay_DrawGameInfo(void)
 {
     M_ResetBarLocations();
-    if (!g_Config.ui.enable_ui) {
-        M_RemoveAmmoText();
-        return;
-    }
 
     Overlay_BarDrawHealth();
     M_BarDrawAir();
@@ -786,10 +797,6 @@ void Overlay_DrawFPSInfo(void)
 
 void Overlay_AddPickup(const GAME_OBJECT_ID object_id)
 {
-    if (!g_Config.ui.enable_ui) {
-        return;
-    }
-
     int32_t grid_x = -1;
     int32_t grid_y = -1;
     for (int i = 0; i < MAX_PICKUPS; i++) {
```

**Comment by lahm86** [comment]
> One more observation: currently, if a player hides the UI immediately after picking up an item and then reveals it again, the pickup animation or sprite continues to be displayed. However, if they perform the actions in reverse order – hiding the UI first, picking up an item, and then showing the UI – the animation does not appear. This seems inconsistent. Here's a suggested patch to address this issue:
> 
> ```diff
> diff --git a/src/tr1/game/overlay.c b/src/tr1/game/overlay.c
> index 46903dd2..aec02ec2 100644
> --- a/src/tr1/game/overlay.c
> +++ b/src/tr1/game/overlay.c
> @@ -251,6 +251,9 @@ static void M_BarGetLocation(
>  
>  void Overlay_BarDraw(BAR_INFO *bar_info, RENDER_SCALE_REF scale_ref)
>  {
> +    if (!g_Config.ui.enable_ui) {
> +        return;
> +    }
>      const RGBA_8888 rgb_bgnd = { 0, 0, 0, 255 };
>      const RGBA_8888 rgb_border = { 53, 53, 53, 255 };
>  
> @@ -320,6 +323,9 @@ static float M_Ease(int32_t cur_frame, int32_t max_frames)
>  
>  static void M_DrawPickup3D(DISPLAY_PICKUP *pu)
>  {
> +    if (!g_Config.ui.enable_ui) {
> +        return;
> +    }
>      int32_t screen_width = Screen_GetResWidth();
>      int32_t screen_height = Screen_GetResHeight();
>      int32_t pickup_width = screen_width / 8;
> @@ -481,6 +487,10 @@ static void M_DrawPickupsSprites(void)
>              continue;
>          }
>  
> +        if (!g_Config.ui.enable_ui) {
> +            continue;
> +        }
> +
>          const int32_t x =
>              Viewport_GetWidth() - sprite_height - sprite_width * pu->grid_x;
>          const int32_t y =
> @@ -587,6 +597,11 @@ static void M_RemoveAmmoText(void)
>  
>  static void M_DrawAmmoInfo(void)
>  {
> +    if (!g_Config.ui.enable_ui) {
> +        M_RemoveAmmoText();
> +        return;
> +    }
> +
>      const double scale = 1.5;
>      const int32_t text_height = 17 * scale;
>      const int32_t text_offset_x = 3;
> @@ -715,10 +730,6 @@ void Overlay_HideGameInfo(void)
>  void Overlay_DrawGameInfo(void)
>  {
>      M_ResetBarLocations();
> -    if (!g_Config.ui.enable_ui) {
> -        M_RemoveAmmoText();
> -        return;
> -    }
>  
>      Overlay_BarDrawHealth();
>      M_BarDrawAir();
> @@ -786,10 +797,6 @@ void Overlay_DrawFPSInfo(void)
>  
>  void Overlay_AddPickup(const GAME_OBJECT_ID object_id)
>  {
> -    if (!g_Config.ui.enable_ui) {
> -        return;
> -    }
> -
>      int32_t grid_x = -1;
>      int32_t grid_y = -1;
>      for (int i = 0; i < MAX_PICKUPS; i++) {
> ```

Thank you, that works nicely 😌

**Review by rr- — APPROVED** [review]

**Pull request merged**